src/group_theory/torsion.lean

```diff
@@ -54,35 +61,106 @@ noncomputable def is_torsion.group [monoid G] (tG : is_torsion G) : group G :=
 lemma exponent_exists.is_torsion (h : exponent_exists G) : is_torsion G := begin
   obtain ⟨n, npos, hn⟩ := h,
   intro g,
   exact (is_of_fin_order_iff_pow_eq_one g).mpr ⟨n, npos, hn g⟩,
 end
 
-/--The group exponent exists for any bounded torsion group. -/
-@[to_additive is_add_torsion.exponent_exists]
+/-- The group exponent exists for any bounded torsion group. -/
+@[to_additive is_add_torsion.exponent_exists
+  "The group exponent exists for any bounded additive torsion group."]
 lemma is_torsion.exponent_exists
   (tG : is_torsion G) (bounded : (set.range (λ g : G, order_of g)).finite) :
   exponent_exists G :=
 exponent_exists_iff_ne_zero.mpr $
   (exponent_ne_zero_iff_range_order_of_finite (λ g, order_of_pos' (tG g))).mpr bounded
 
-/--Finite groups are torsion groups.-/
-@[to_additive is_add_torsion_of_fintype]
+/-- Finite groups are torsion groups. -/
+@[to_additive is_add_torsion_of_fintype "Finite additive groups are additive torsion groups."]
 lemma is_torsion_of_fintype [fintype G] : is_torsion G :=
 exponent_exists.is_torsion $ exponent_exists_iff_ne_zero.mpr exponent_ne_zero_of_fintype
+
+end group
+
+
+section comm_monoid
+
+variables (G) [comm_monoid G]
+
+namespace comm_monoid
+
+/--
+The torsion submonoid of a commutative monoid.
+
+(Note that by `monoid.is_torsion.group` torsion monoids are truthfully groups.)
+-/
+@[to_additive add_torsion "The torsion submonoid of an additive commutative monoid."]
+def torsion : submonoid G :=
+{ carrier := {x | is_of_fin_order x},
+  one_mem' := is_of_fin_order_one,
+  mul_mem' := λ _ _ hx hy, hx.mul hy }
+
+variable {G}
+
+/-- Torsion submonoids are torsion. -/
+@[to_additive "Additive torsion submonoids are additively torsion."]
+lemma torsion.is_torsion : is_torsion $ torsion G :=
+λ ⟨_, n, npos, hn⟩,
+  ⟨n, npos, subtype.ext $
+    by rw [mul_left_iterate, _root_.mul_one, submonoid.coe_pow,
+           subtype.coe_mk, submonoid.coe_one, (is_periodic_pt_mul_iff_pow_eq_one _).mp hn]⟩
+
+end comm_monoid
+
+open comm_monoid (torsion)
+
+namespace monoid.is_torsion
+
+variable {G}
+
+/-- The torsion submonoid of a torsion monoid is `⊤`. -/
+@[simp, to_additive "The additive torsion submonoid of an additive torsion monoid is `⊤`."]
+lemma torsion_eq_top (tG : is_torsion G) : torsion G = ⊤ := by ext; tauto
+
+/-- A torsion monoid is isomorphic to its torsion submonoid. -/
+@[to_additive "An additive torsion monoid is isomorphic to its torsion submonoid."]
+def torsion_mul_equiv (tG : is_torsion G) : torsion G ≃* G :=
+by rw tG.torsion_eq_top; exact submonoid.top_equiv
+
+end monoid.is_torsion
+
+/-- Torsion submonoids of a torsion submonoid are isomorphic to the submonoid. -/
+@[simp, to_additive add_comm_monoid.torsion.of_torsion
+  "Additive torsion submonoids of an additive torsion submonoid are isomorphic to the submonoid."]
+def torsion.of_torsion : (torsion (torsion G)) ≃* (torsion G) :=
+monoid.is_torsion.torsion_mul_equiv comm_monoid.torsion.is_torsion
```

Comment:
```suggestion
comm_monoid.torsion.is_torsion.torsion_mul_equiv
```

Reply:
Lean kept giving strange type errors here that I don't understand. With full dot notation (as you suggested as well), it says:

```
▶ expected type:
G : Type u_1
_inst_1 : comm_monoid G
⊢ Type (u_1+1)

▶ 195:1-195:2: error:
type mismatch at application
  comm_monoid.torsion.is_torsion.torsion_mul_equiv
term
  comm_monoid.torsion.is_torsion
has type
  is_torsion ↥(torsion ?m_1)
but is expected to have type
  is_torsion ?m_1
Additional information:
/Users/julian/Development/mathlib/src/group_theory/torsion.lean: context: switched to simple application elaboration procedure because failed to use expected type to elaborate it, error message
  type mismatch at application
    comm_monoid.torsion.is_torsion.torsion_mul_equiv
  term
    comm_monoid.torsion.is_torsion
  has type
    is_torsion ↥(torsion ?m_1)
  but is expected to have type
    is_torsion ?m_1
```